### PR TITLE
chore: replace ResolveRequestContextBase and introduce IResolveRequest as a replacement

### PR DESCRIPTION
**Pull request by alsami**
As discussed in #1164, this PR removes the class `ResolveRequestContextBase` and introduces the interface `IResolveRequestContext` deriving from `IComponentContext`

Changes:
* Delete `ResolveRequestContextBase`
* Make `ResolveRequestContext` implement `IResolveRequestContext`
* Replace all references to `ResolveRequestContextBase` with `IResolveRequestContext`

### CODE REVIEW

**Comment by alsami** [comment]
Went surprisingly well, kindly requesting a review @tillig @alistairjevans

File: src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs [hunk 1]
```diff
@@ -23,40 +23,107 @@
...
+        public object ResolveComponent(ResolveRequest request) =>
+            Operation.GetOrCreateInstance(ActivationScope, request);
+
+
+        /// <inheritdoc/>
+        public void CompleteRequest()
         {
-            // Let the base class raise events.
-            CompleteRequest();
+            EventHandler<ResolveRequestCompletingEventArgs>? handler = RequestCompleting;
```
Comment (alsami) [inline src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:125]: Is there a reason, why this is copied to a local variable again? Why not just `RequestCompleting?.Invoke(this, new ...)`;

Comment (alistairjevans) [reply src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:124]: Something in my brain is telling me that it's possible for that RequestCompleting handler to be changed from inside the event handler itself? Maybe? Or it's a GC thing? I can't entirely remember tbh.

File: src/Autofac/Core/Resolving/Pipeline/IResolveRequestContext.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        /// <summary>
+        /// Change the set of parameters being used in the processing of this request.
+        /// </summary>
+        /// <param name="newParameters">The new set of parameters.</param>
+        void ChangeParameters(IEnumerable<Parameter> newParameters);
+
+        /// <summary>
+        ///     Complete the request, raising any appropriate events.
+        /// </summary>
+        public void CompleteRequest();
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/Pipeline/IResolveRequestContext.cs:95]: `CompleteRequest` should really only be called by the `ResolveOperation` itself, not by a consumer of the interface...is there a way to do it so this method stays only in the internal implementation?

File: src/Autofac/Core/Resolving/Pipeline/IResolveRequestContext.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        void ChangeScope(ISharingLifetimeScope newScope);
+
+        /// <summary>
+        /// Change the set of parameters being used in the processing of this request.
+        /// </summary>
+        /// <param name="newParameters">The new set of parameters.</param>
+        void ChangeParameters(IEnumerable<Parameter> newParameters);
+
+        /// <summary>
+        ///     Complete the request, raising any appropriate events.
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/Pipeline/IResolveRequestContext.cs:93]: Leading spaces.

File: src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs [hunk 4]
```diff
@@ -23,40 +23,106 @@
...
+using Autofac.Features.Decorators;
 
 namespace Autofac.Core.Resolving.Pipeline
 {
     /// <summary>
-    /// Context area for a resolve request.
+    ///     Context area for a resolve request.
     /// </summary>
-    internal sealed class ResolveRequestContext : ResolveRequestContextBase
+    internal class ResolveRequestContext : IResolveRequestContext
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:37]: Can this stay sealed if we're exposing an interface? Not sure why we'd need to derive from it now.

File: src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs [hunk 5]
```diff
@@ -23,40 +23,106 @@
...
     {
+        private readonly ResolveRequest _resolveRequest;
+        private object? _instance;
+
         /// <summary>
-        /// Initializes a new instance of the <see cref="ResolveRequestContext"/> class.
+        ///     Initializes a new instance of the <see cref="ResolveRequestContext" /> class.
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:43]: Leading space

File: src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs [hunk 6]
```diff
@@ -23,40 +23,106 @@
...
         /// </summary>
         /// <param name="owningOperation">The owning resolve operation.</param>
         /// <param name="request">The initiating resolve request.</param>
         /// <param name="scope">The lifetime scope.</param>
         /// <param name="diagnosticSource">
-        /// The <see cref="System.Diagnostics.DiagnosticListener"/> to which trace events should be written.
+        ///     The <see cref="System.Diagnostics.DiagnosticListener" /> to which trace events should be written.
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:49]: Leading space.

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 7]
```diff
@@ -34,169 +34,102 @@
 namespace Autofac.Core.Resolving
 {
     /// <summary>
-    /// Defines the base properties and behaviour of a resolve operation.
+    ///     Defines the base properties and behaviour of a resolve operation.
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:37]: Leading space

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 8]
```diff
@@ -34,169 +34,102 @@
...
-        private readonly List<ResolveRequestContext> _successfulRequests = new List<ResolveRequestContext>(SuccessListInitialCapacity);
+
+        private readonly List<IResolveRequestContext> _successfulRequests =
+            new List<IResolveRequestContext>(SuccessListInitialCapacity);
+
         private int _nextCompleteSuccessfulRequestStartPos;
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="ResolveOperationBase"/> class.
+        ///     Initializes a new instance of the <see cref="ResolveOperationBase" /> class.
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:50]: Leading space

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 9]
```diff
@@ -34,169 +34,102 @@
...
         /// </summary>
-        /// <param name="mostNestedLifetimeScope">The most nested scope in which to begin the operation. The operation
-        /// can move upward to less nested scopes as components with wider sharing scopes are activated.</param>
+        /// <param name="mostNestedLifetimeScope">
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:52]: Is something odd going on with the line breaks here? Might just be a github thing.

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 10]
```diff
@@ -34,169 +34,102 @@
...
+            ISharingLifetimeScope mostNestedLifetimeScope,
+            DiagnosticListener diagnosticSource)
         {
             CurrentScope = mostNestedLifetimeScope ?? throw new ArgumentNullException(nameof(mostNestedLifetimeScope));
             DiagnosticSource = diagnosticSource ?? throw new ArgumentNullException(nameof(diagnosticSource));
         }
 
         /// <summary>
-        /// Gets the active resolve request.
+        ///     Gets the active resolve request.
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:68]: Loads of leading spacing in this file. Indentation isn't really needed.

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 11]
```diff
@@ -34,169 +34,102 @@
...
-        }
-
         /// <inheritdoc />
         public object GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, ResolveRequest request)
         {
-            if (request is null)
-            {
-                throw new ArgumentNullException(nameof(request));
-            }
+            if (request is null) throw new ArgumentNullException(nameof(request));
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:118]: We're switching to always use curly brances, so best leave that as it was.

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 12]
```diff
@@ -34,169 +34,102 @@
...
 
             if (_ended) throw new ObjectDisposedException(ResolveOperationResources.TemporaryContextDisposed, innerException: null);
 
             // Create a new request context.
             var requestContext = new ResolveRequestContext(this, request, currentOperationScope, DiagnosticSource);
 
             // Raise our request-beginning event.
-            var handler = ResolveRequestBeginning;
+            EventHandler<ResolveRequestBeginningEventArgs>? handler = ResolveRequestBeginning;
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:126]: Why the change away from var?

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 13]
```diff
@@ -247,45 +180,115 @@ public object GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, R
...
         {
-            if (_ended) return;
+            if (_ended)
+            {
+                return;
+            }
 
             _ended = true;
-            var handler = CurrentOperationEnding;
+            EventHandler<ResolveOperationEndingEventArgs>? handler = CurrentOperationEnding;
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:291]: Don't need to switch from var.

File: src/Autofac/IComponentContext.cs [hunk 14]
```diff
@@ -38,7 +38,7 @@ public interface IComponentContext
         /// <summary>
         /// Gets the associated services with the components that provide them.
         /// </summary>
-        IComponentRegistry ComponentRegistry { get; }
+        public IComponentRegistry ComponentRegistry { get; }
```
Comment (alistairjevans) [inline src/Autofac/IComponentContext.cs:41]: Don't need public markers on interfaces.

File: src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs (hunk 4 above)
Comment (alsami) [reply src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:37]: `PipelineBuilderTests`  class contains a `MockPiplineRequest` which derived from `ResolveRequestContextBase`. That's why I've unsealed it. Should we change that?

Comment (alistairjevans) [reply src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:37]: You could argue that it should now just implement the interface? 

This feeds into the argument for revisiting the pipeline tests we have to use the new approach to mocking the context.

**Comment by alsami** [comment]
> Some minor blips to take a look at.

I hope I've addressed everything, except of the sealed class thingy, waiting for your response there.

> Also, I think as part of these changes we need to get the tests updated that previously had to jump through hoops to invoke/test middleware. Partly just to verify that these changes are the right ones.

If you don't mind, I will do that in a second run. I'd collect the remaining Tasks in #1164

File: src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs (hunk 4 above)
Comment (alsami) [reply src/Autofac/Core/Resolving/Pipeline/ResolveRequestContext.cs:37]: Agree, would like to do that in a separate PR if you don't mind.

File: src/Autofac/IComponentContext.cs [hunk 15]
```diff
@@ -49,6 +49,6 @@ public interface IComponentContext
         /// </returns>
         /// <exception cref="ComponentNotRegisteredException"/>
         /// <exception cref="DependencyResolutionException"/>
-        object ResolveComponent(ResolveRequest request);
+        public object ResolveComponent(ResolveRequest request);
```
Comment (alistairjevans) [inline src/Autofac/IComponentContext.cs:52]: This modifier isn't needed either.

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 16]
```diff
@@ -246,43 +184,113 @@ public object GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, R
...
-                completed[i].Complete();
+                completed[i].CompleteRequest();
             }
         }
 
-        private void ResetSuccessfulRequests()
-        {
-            _nextCompleteSuccessfulRequestStartPos = _successfulRequests.Count;
-        }
+        private void ResetSuccessfulRequests() => _nextCompleteSuccessfulRequestStartPos = _successfulRequests.Count;
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:286]: Not sure how I feel about expression-based methods that don't return things...style preference I think.

**Comment by alsami** [comment]
> Little bits left behind, but yes, once those are done let's get this merged and we can update the tests separately.

Should be good now, I hope.

File: src/Autofac/Core/Resolving/ResolveOperationBase.cs [hunk 17]
```diff
@@ -246,32 +184,102 @@ public object GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, R
...
+            return result;
+        }
+
         /// <summary>
         /// Basic pipeline invocation steps used when retrieving an instance. Isolated
         /// to enable it to be optionally surrounded with diagnostics.
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void InvokePipeline(ResolveRequest request, ResolveRequestContext requestContext)
+        private void InvokePipeline(ResolveRequest request, ResolveRequestContext requestContextContext)
```
Comment (alistairjevans) [inline src/Autofac/Core/Resolving/ResolveOperationBase.cs:262]: Sorry, I missed this naming issue first time out...

Comment (alsami) [reply src/Autofac/Core/Resolving/ResolveOperationBase.cs]: Search & Replace does not always work as desired :roll_eyes:

**Comment by alsami** [comment]
> Out of interest, how come you always force-push your changes? It can make it harder to see what has changed since the last review.

I believe that one feature/fix/chore/generic-change should always contain exactly one commit. Makes browsing and understanding the history easier.

If it's better to not do that, I can create separate commits until you say it's fine, then squash the changes.

**Comment by alistairjevans** [comment]
I'd usually prefer to start out with all the changes, then squash (but only if we feel it's warranted). Squashing at the end of a PR makes the review process easy if there's a few discussions that take place.

**Comment by alsami** [comment]
> I'd usually prefer to start out with all the changes, then squash (but only if we feel it's warranted). Squashing at the end of a PR makes the review process easy if there's a few discussions that take place.

Oki, will do my best, really got to be a habit of mine, have dedicated git-aliases to do it with one command basically :smile:

**Comment by tillig** [comment]
I'm late to the party, but I don't have a strong personal preference about squash or not squash. I did just try to do that with the DiagnosticSource changes I made, but after I had to pull a couple of things from the v6 codeline and there was a bunch of churn, it ended up turning into a nightmare of Git thinking that I made every change including changes I didn't make and I was like... "yeah, I'm not fighting this." There are definitely folks better at Git-fu than I am, that's for sure.

**Comment by alsami** [comment]
Yeah, squashing and rebasing require some practice, I ran into all these problems as well.

It's worth it though, basically it looks like that everything was developed and pushed on the trunk directly. Very helpful when working on a bigger project & team.

You can even use "Rebase and Merge" on github, which will not create a merge-commit and basically just fast-forwards.

**Comment by alistairjevans** [comment]
I can highly recommend the Fork git client, which has a built-in interactive rebase/squash feature, that makes all this much more intuitive.
